cmd/catalyst-node/catalyst-node.go

```diff
@@ -478,11 +497,84 @@ func redirectHandler(redirectPrefixes []string, nodeHost string) http.Handler {
 
 		rPath := fmt.Sprintf(pathTmpl, fullPlaybackID)
 		rURL := fmt.Sprintf("%s://%s%s", protocol(r), bestNode, rPath)
+		rURL, err = resolveNodeURL(rURL)
+		if err != nil {
+			glog.Errorf("failed to resolve node URL playbackID=%s err=%s", playbackID, err)
+			w.WriteHeader(http.StatusInternalServerError)
+			return
+		}
 		glog.V(6).Infof("generated redirect url=%s", rURL)
 		http.Redirect(w, r, rURL, http.StatusFound)
 	})
 }
 
+func streamSourceHandler() http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		b, err := io.ReadAll(r.Body)
+		if err != nil {
+			glog.Errorf("error handling STREAM_SOURCE body=%s", err)
+			w.Write([]byte("push://"))
+			return
+		}
+		streamName := string(b)
+		glog.V(7).Infof("got mist STREAM_SOURCE request=%s", streamName)
+		dtscURL, err := queryMistForClosestNodeSource(streamName, "0", "0", "", true)
+		if err != nil {
+			glog.Errorf("error querying mist for STREAM_SOURCE: %s", err)
+			w.Write([]byte("push://"))
+			return
+		}
+		outURL, err := resolveNodeURL(dtscURL)
+		if err != nil {
+			glog.Errorf("error finding STREAM_SOURCE: %s", err)
+			w.Write([]byte("push://"))
+			return
+		}
+		glog.V(7).Infof("replying to Mist STREAM_SOURCE request=%s response=%s", streamName, outURL)
+		w.Write([]byte(outURL))
+	})
+}
+
+// Given a dtsc:// or https:// url, resolve the proper address of the node via serf tags
+func resolveNodeURL(streamURL string) (string, error) {
+	u, err := url.Parse(streamURL)
+	if err != nil {
+		return "", err
+	}
+	nodeName := u.Host
+	protocol := u.Scheme
+
+	member, err := getSerfMember(nodeName)
+	addr, has := member.Tags[protocol]
+	if !has {
+		glog.V(7).Infof("no tag found, not tag resolving protocol=%s nodeName=%s", protocol, nodeName)
```

Comment:
Do we plan to tag all members with protocols?

Reply:
Yep!